### PR TITLE
fix(sharding): Make sharding async migration more reliable

### LINKED ISSUE #8920: Async migration 0004: Transaction failed (Node exists)
## Bug description

I ran the 0004 async migration on benchmarking server. This failed with:

```
Exception was thrown while running operation 33 : Code: 999. DB::Exception: Transaction failed (Node exists): Op #12, path: /clickhouse/tables/noshard/posthog.person_distinct_id2/blocks/all_replace_from_687FBECF78B9DCF494A054EDBA6BEF17. Stack trace: 0. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) @ 0x8b6cbba in /usr/bin/clickhouse 1. Coordination::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, Coordination::Error, int) @ 0x107c1635 in /usr/bin/clickhouse 2. Coordination::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, Coordination::Error) @ 0x107c18c2 in /usr/bin/clickhouse 3. zkutil::KeeperMultiException::KeeperMultiException(Coordination::Error, std::__1::vector<std::__1::shared_ptr<Coordination::Request>, std::__1::allocator<std::__1::shared_ptr<Coordination::Request> > > const&, std::__1::vector<std::__1::shared_ptr<Coordination::Response>, std::__1::allocator<std::__1::shared_ptr<Coordination::Response> > > const&) @ 0x107d4284 in /usr/bin/clickhouse 4. zkutil::KeeperMultiException::check(Coordination::Error, std::__1::vector<std::__1::shared_ptr<Coordination::Request>, std::__1::allocator<std::__1::shared_ptr<Coordination::Request> > > const&, std::__1::vector<std::__1::shared_ptr<Coordination::Response>, std::__1::allocator<std::__1::shared_ptr<Coordination::Response> > > const&) @ 0x107c9aaf in /usr/bin/clickhouse 5. DB::StorageReplicatedMergeTree::replacePartitionFrom(std::__1::shared_ptr<DB::IStorage> const&, std::__1::shared_ptr<DB::IAST> const&, bool, std::__1::shared_ptr<DB::Context>) @ 0xff5b1b2 in /usr/bin/clickhouse 6. DB::MergeTreeData::alterPartition(std::__1::shared_ptr<DB::StorageInMemoryMetadata const> const&, std::__1::vector<DB::PartitionCommand, std::__1::allocator<DB::PartitionCommand> > const&, std::__1::shared_ptr<DB::Context>) @ 0x1013e9ec in /usr/bin/clickhouse 7. DB::InterpreterAlterQuery::execute() @ 0xf6c5ee5 in /usr/bin/clickhouse 8. ? @ 0xfc0a4a1 in /usr/bin/clickhouse 9. DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum, bool) @ 0xfc08b23 in /usr/bin/clickhouse 10. DB::TCPHandler::runImpl() @ 0x1042e932 in /usr/bin/clickhouse 11. DB::TCPHandler::run() @ 0x10441839 in /usr/bin/clickhouse 12. Poco::Net::TCPServerConnection::start() @ 0x12a3fd4f in /usr/bin/clickhouse 13. Poco::Net::TCPServerDispatcher::run() @ 0x12a417da in /usr/bin/clickhouse 14. Poco::PooledThread::run() @ 0x12b7ab39 in /usr/bin/clickhouse 15. Poco::ThreadImpl::runnableEntry(void*) @ 0x12b76b2a in /usr/bin/clickhouse 16. start_thread @ 0x9609 in /usr/lib/x86_64-linux-gnu/libpthread-2.31.so 17. __clone @ 0x122293 in /usr/lib/x86_64-linux-gnu/libc-2.31.so
```

## Expected behavior

No errors

## How to reproduce

1.
2.
3.

## Environment

- [ ] PostHog Cloud
- [x] self-hosted PostHog (ClickHouse-based), version/commit: latest
- [ ] self-hosted PostHog (Postgres-based, legacy), version/commit: _please provide_

## Additional context

prerequisite for https://github.com/PostHog/posthog/issues/8652

Note that the 0003 migration was run just before that on this table - might be the cause?

### PR DESCRIPTION
**Pull request by macobo**
## Problem

Sharding async migration needs to be more reliable.

Closes https://github.com/PostHog/posthog/issues/8920, part of https://github.com/PostHog/posthog/issues/8652

Follow-up to https://github.com/PostHog/posthog/pull/8982, depends on https://github.com/PostHog/posthog/pull/9010

## Changes

This change:
- Makes the rollbacks always work by
    - Fixing some operations from before
    - Creating/deleting materialized views correctly
    - Ensuring zookeeper paths are unique
- Handles an edge case around moving tables by retrying

## How did you test this code?

Deployed onto benchmarking server, unit tests.

### CODE REVIEW

File: posthog/async_migrations/migrations/0004_replicated_schema.py [hunk 1]
```diff
@@ -108,9 +120,12 @@ def precheck(self):
...
     def operations(self):
-        TABLE_MIGRATION_OPERATIONS = [
-            operation for table in self.tables_to_migrate() for operation in self.replicated_table_operations(table)
-        ]
+        TABLE_MIGRATION_OPERATIONS = list(
+            flatten([list(self.replicated_table_operations(table)) for table in self.tables_to_migrate()])
+        )
+        RE_ENABLE_INGESTION_OPERATIONS = list(
+            flatten([list(self.finalize_table_operations(table)) for table in self.tables_to_migrate()])
+        )
```
Comment (tiina303) [inline posthog/async_migrations/migrations/0004_replicated_schema.py:128]: why? I find this harder to read

File: posthog/async_migrations/migrations/0004_replicated_schema.py [hunk 2]
```diff
@@ -219,35 +236,46 @@ def move_partitions(self, from_table: str, to_table: str):
...
+                        "Moving partitions between tables", from_table=from_table, to_table=to_table, id=partition
+                    )
+                    # :KLUDGE: Partition IDs are special and cannot be passed as arguments
+                    sync_execute(f"ALTER TABLE {to_table} REPLACE PARTITION {partition} FROM {from_table}")
+                    sync_execute(f"ALTER TABLE {from_table} DROP PARTITION {partition}")
+                break
+            # :KLUDGE: Logic to retry moving parts for person_distinct_id2 table
+            except ServerException as err:
+                error_code = lookup_error_code(err)
+                if error_code != "KEEPER_EXCEPTION" or retry > self.MOVE_PARTS_RETRIES:
```
Comment (tiina303) [inline posthog/async_migrations/migrations/0004_replicated_schema.py:258]: This means that we're going to do `self.MOVE_PARTS_RETRIES + 1` retries, which I find slightly confusing to read/reason about. Consider
```                
  retries_left = self.MOVE_PARTS_RETRIES
  while True:
    ...
    if error_code != "KEEPER_EXCEPTION" or not retries_left:  # or retries_left <= 0
       ...
    else:
      retries_left -= 1
```

File: posthog/async_migrations/migrations/0004_replicated_schema.py [hunk 3]
```diff
@@ -219,35 +236,46 @@ def move_partitions(self, from_table: str, to_table: str):
...
+                    retry += 1
+                    logger.warning(
+                        "Moving part failed due to (potentially) sporatic zookeeper exception. Retrying...",
+                        from_table=from_table,
+                        to_table=to_table,
+                        retry=retry,
+                    )
+
+    def rename_tables(self, rename_1, rename_2, verification_table=None):
         # :KLUDGE: Due to how async migrations rollback works, we need to check whether backup table exists even if the rename failed
```
Comment (tiina303) [inline posthog/async_migrations/migrations/0004_replicated_schema.py:270]:
```suggestion
        # :KLUDGE: Due to how async migrations rollback works, we need to check whether verification table exists even if the rename failed
```

**Review by tiina303 — APPROVED** [review]
LGTM but maybe let Yakko 👀  too

File: posthog/async_migrations/migrations/0004_replicated_schema.py (hunk 1 above)
Comment (macobo) [reply posthog/async_migrations/migrations/0004_replicated_schema.py:128]: Why what?

File: posthog/async_migrations/migrations/0004_replicated_schema.py [hunk 4]
```diff
@@ -51,15 +56,18 @@
...
 
 
 @dataclass(frozen=True)
 class TableMigrationData:
     name: str
     new_table_engine: MergeTreeEngine
-    materialized_view_name: Optional[str]
-    create_materialized_view: Optional[str]
+    kafka_table_name: Optional[str]
+    create_kafka_table: Optional[str]
```
Comment (yakkomajuri) [inline posthog/async_migrations/migrations/0004_replicated_schema.py:70]: nit: this has a name that implies a function to me

File: posthog/async_migrations/migrations/0004_replicated_schema.py [hunk 5]
```diff
@@ -219,35 +236,46 @@ def move_partitions(self, from_table: str, to_table: str):
...
+                break
+            # :KLUDGE: Logic to retry moving parts for person_distinct_id2 table
+            except ServerException as err:
+                error_code = lookup_error_code(err)
+                if error_code != "KEEPER_EXCEPTION" or retry > self.MOVE_PARTS_RETRIES:
+                    raise err
+                else:
+                    retry += 1
+                    logger.warning(
+                        "Moving part failed due to (potentially) sporatic zookeeper exception. Retrying...",
```
Comment (yakkomajuri) [inline posthog/async_migrations/migrations/0004_replicated_schema.py:263]:
```suggestion
                        "Moving part failed. This is potentially due to a sporadic zookeeper exception. Retrying...",
```

File: posthog/async_migrations/migrations/0004_replicated_schema.py [hunk 6]
```diff
@@ -147,27 +163,33 @@ def replicated_table_operations(self, table: TableMigrationData):
...
 
-        if table.materialized_view_name is not None:
+    def finalize_table_operations(self, table: TableMigrationData):
+        # NOTE: Relies on IF NOT EXISTS on the query
+        if isinstance(table, ShardedTableMigrationData):
+            for table_name, create_table_query in table.extra_tables.items():
+                yield AsyncMigrationOperationSQL(sql=create_table_query, rollback=f"DROP TABLE IF EXISTS {table_name}")
+
+        if isinstance(table, ShardedTableMigrationData) and table.materialized_view_name is not None:
+            yield AsyncMigrationOperationSQL(sql=f"DROP TABLE IF EXISTS {table.materialized_view_name}", rollback=None)
```
Comment (yakkomajuri) [inline posthog/async_migrations/migrations/0004_replicated_schema.py:182]: nit: could we use `DROP VIEW` for materialized views? Literally won't change a thing, just helps with readability amongst all the drop statements

**Review by yakkomajuri — APPROVED** [review]
the changes look ok to me, but I haven't run this and don't have amazing context on all the steps. let me know if you'd like a very thorough review of functionality and all the sql statements and I can run this locally and inspect statements one by one.

File: posthog/async_migrations/migrations/0004_replicated_schema.py (hunk 6 above)
Comment (macobo) [reply posthog/async_migrations/migrations/0004_replicated_schema.py:182]: Migrations also use `DROP TABLE` though.

Comment (yakkomajuri) [reply posthog/async_migrations/migrations/0004_replicated_schema.py:182]: yeah that's true

**Pull request merged**